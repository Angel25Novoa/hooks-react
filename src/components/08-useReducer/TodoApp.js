import React, {useEffect, useReducer} from 'react'
import todoReducer from './todoReducer'
import TodoForm from './TodoForm'

import './styles.css'
import TodoList from './TodoList'

const initialState = [{
  id: new Date().getTime(),
  desc: 'Learn Hooks',
  done: false
}]

const init = () => {
  return localStorage.getItem('todos') ? JSON.parse(localStorage.getItem('todos')) : []
}

const TodoApp = () => {

  const [ todos, dispatch ] = useReducer(todoReducer, initialState, init)

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])

  const handleDelete = (todoId) => {

  console.log(todoId)
    // crear la acciÃ³n
    const removeAction = {
      type: 'remove',
      payload: todoId
    }
    //dispatch(action)
    dispatch(removeAction)
  }

  const handleToggle = (todoId) => {
    const toggleAction = {
      type: 'toggle',
      payload: todoId
    }
    dispatch(toggleAction)
  }

  const handleAddTodo = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo
    })
  }

  return (
    <div>
      <h1>Todo App ({todos.length})</h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList 
            todos={todos} 
            handleDelete={handleDelete} 
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <TodoForm  
            handleAddTodo={handleAddTodo}
          />
        </div>
      </div>
    </div>
  )
}

export default TodoApp