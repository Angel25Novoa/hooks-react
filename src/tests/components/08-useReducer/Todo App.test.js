import React from 'react'
import {shallow, mount} from 'enzyme'
import TodoApp from '../../../components/08-useReducer/TodoApp'
// import TodoForm from '../../../components/08-useReducer/TodoForm'
import { demoState } from '../../fixtures/demoTodos'
import {act} from '@testing-library/react'

describe((`TodoApp tests`), () => {

  const wrapper = shallow(<TodoApp/>)
  Storage.prototype.setItem = jest.fn()

  test((`TodoApp component should render without errors`), () => {
    expect(wrapper).toMatchSnapshot()
  })
  test('should add a ToDo', () => {
    const wrapper = mount(<TodoApp/>)
    act(() => {
      wrapper.find('TodoForm').prop('handleAddTodo')(demoState[0])
      wrapper.find('TodoForm').prop('handleAddTodo')(demoState[1])
    })
    expect(wrapper.find('h1').text().trim()).toBe('Todo App (2)')
    expect(localStorage.setItem).toHaveBeenCalledTimes(2)
    expect(localStorage.setItem).toHaveBeenCalledWith('todos', JSON.stringify(demoState))
  })
  test('should delete a ToDo', () => {
    wrapper.find('TodoForm').prop('handleAddTodo')(demoState[0])
    expect(wrapper.find('h1').text().trim()).toBe('Todo App (1)')
    wrapper.find('TodoList').prop('handleDelete')(demoState[0].id)
    expect(wrapper.find('h1').text().trim()).toBe('Todo App (0)')
  })
})
