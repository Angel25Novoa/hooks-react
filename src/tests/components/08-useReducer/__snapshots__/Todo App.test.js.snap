// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp tests TodoApp component should render without errors 1`] = `
<div>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoForm
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
